diff --git a/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/ClusterTopicRouteService.java b/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/ClusterTopicRouteService.java
index fb97002df..84252f8b8 100644
--- a/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/ClusterTopicRouteService.java
+++ b/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/ClusterTopicRouteService.java
@@ -17,7 +17,6 @@
 package org.apache.rocketmq.proxy.service.route;
 
 import java.util.List;
-import org.apache.rocketmq.common.MixAll;
 import org.apache.rocketmq.common.message.MessageQueue;
 import org.apache.rocketmq.proxy.common.Address;
 import org.apache.rocketmq.client.impl.mqclient.MQClientAPIFactory;
@@ -59,11 +58,8 @@ public class ClusterTopicRouteService extends TopicRouteService {
 
     @Override
     public String getBrokerAddr(ProxyContext ctx, String brokerName) throws Exception {
-        List<BrokerData> brokerDataList = getAllMessageQueueView(ctx, brokerName).getTopicRouteData().getBrokerDatas();
-        if (brokerDataList.isEmpty()) {
-            return null;
-        }
-        return brokerDataList.get(0).getBrokerAddrs().get(MixAll.MASTER_ID);
+        TopicRouteWrapper topicRouteWrapper = getAllMessageQueueView(ctx, brokerName).getTopicRouteWrapper();
+        return topicRouteWrapper.getMasterAddr(brokerName);
     }
 
     @Override
diff --git a/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/MessageQueueView.java b/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/MessageQueueView.java
index b3a6b9e4b..fe5387cfd 100644
--- a/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/MessageQueueView.java
+++ b/proxy/src/main/java/org/apache/rocketmq/proxy/service/route/MessageQueueView.java
@@ -37,6 +37,10 @@ public class MessageQueueView {
         return topicRouteWrapper.getTopicRouteData();
     }
 
+    public TopicRouteWrapper getTopicRouteWrapper() {
+        return topicRouteWrapper;
+    }
+
     public String getTopicName() {
         return topicRouteWrapper.getTopicName();
     }
