diff --git a/src/main/java/org/assertj/core/internal/Maps.java b/src/main/java/org/assertj/core/internal/Maps.java
index e98e5a4e7..2aa78c595 100644
--- a/src/main/java/org/assertj/core/internal/Maps.java
+++ b/src/main/java/org/assertj/core/internal/Maps.java
@@ -419,6 +419,8 @@ public class Maps {
 
   @SuppressWarnings("unchecked")
   private static <K, V> Map<K, V> clone(Map<K, V> map) throws NoSuchMethodException {
+    if (isMultiValueMapAdapterInstance(map)) throw new IllegalArgumentException("Cannot clone MultiValueMapAdapter");
+
     try {
       if (map instanceof Cloneable) {
         return (Map<K, V>) map.getClass().getMethod("clone").invoke(map);
@@ -438,6 +440,19 @@ public class Maps {
     }
   }
 
+  private static boolean isMultiValueMapAdapterInstance(Map<?, ?> map) {
+    return isInstanceOf(map, "org.springframework.util.MultiValueMapAdapter");
+  }
+
+  private static boolean isInstanceOf(Object object, String className) {
+    try {
+      Class<?> type = Class.forName(className);
+      return type.isInstance(object);
+    } catch (ClassNotFoundException e) {
+      return false;
+    }
+  }
+
   public <K, V> void assertContainsValue(AssertionInfo info, Map<K, V> actual, V value) {
     assertNotNull(info, actual);
     if (!containsValue(actual, value)) throw failures.failure(info, shouldContainValue(actual, value));
