diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheck.java b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheck.java
index 14bf018b0..77f825a1a 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanExpressionCheck.java
@@ -23,11 +23,13 @@ import com.puppycrawl.tools.checkstyle.StatelessCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
  * <p>
  * Checks for over-complicated boolean expressions. Currently finds code like
  * {@code if (b == true)}, {@code b || true}, {@code !false},
+ * {@code boolean a = q > 12 ? true : false},
  * etc.
  * </p>
  * <p>
@@ -59,6 +61,9 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  *     e = (a || false) ? c : d; // violation, can be simplified to a
  *     e = (a &amp;&amp; b) ? c : d;     // OK
  *
+ *     int s = 12;
+ *     boolean m = s &gt; 1 ? true : false; // violation, can be simplified to s &gt; 1
+ *     boolean f = c == null ? false : c.someMethod(); // OK
  *  }
  *
  * }
@@ -113,6 +118,15 @@ public class SimplifyBooleanExpressionCheck
             case TokenTypes.LAND:
                 log(parent, MSG_KEY);
                 break;
+            case TokenTypes.QUESTION:
+                final DetailAST nextSibling = ast.getNextSibling();
+                if (TokenUtil.isBooleanLiteralType(parent.getFirstChild().getType())
+                        || nextSibling != null
+                        && TokenUtil.isBooleanLiteralType(
+                        nextSibling.getNextSibling().getType())) {
+                    log(parent, MSG_KEY);
+                }
+                break;
             default:
                 break;
         }
diff --git a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanExpressionCheck.xml b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanExpressionCheck.xml
index af45aa78c..07a62ed3e 100644
--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanExpressionCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanExpressionCheck.xml
@@ -7,6 +7,7 @@
          <description>&lt;p&gt;
  Checks for over-complicated boolean expressions. Currently finds code like
  {@code if (b == true)}, {@code b || true}, {@code !false},
+ {@code boolean a = q &gt; 12 ? true : false},
  etc.
  &lt;/p&gt;
  &lt;p&gt;
