diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java
index dac05e893..d089a1698 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java
@@ -423,12 +423,7 @@ public class OperatorWrapCheck
             result = adjustParens(rightNode, DetailAST::getPreviousSibling);
         }
 
-        // The ARRAY_INIT AST is confusing. It should be
-        // ARRAY_INIT
-        // |--LCURLY (first child)
-        // `--RCURLY (last child)
-        // but there is no LCURLY for ARRAY_INIT
-        if (result.getType() != TokenTypes.ARRAY_INIT) {
+        if (!TokenUtil.isOfType(result, TokenTypes.ARRAY_INIT, TokenTypes.ANNOTATION_ARRAY_INIT)) {
             while (result.getFirstChild() != null) {
                 result = result.getFirstChild();
             }
