diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
index 81819783b..f8ae785b5 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
@@ -111,7 +111,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <li>
  * Property {@code memberModifiers} - Control whether to check only methods and fields with any
  * of the specified modifiers.
- * This property does not affect method calls nor method references.
+ * This property does not affect method calls nor method references nor record components.
  * Type is {@code java.lang.String[]}.
  * Validation type is {@code tokenTypesSet}.
  * Default value is {@code ""}.
@@ -380,7 +380,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
     private final Set<String> ignoredMethodNames = new HashSet<>();
     /**
      * Control whether to check only methods and fields with any of the specified modifiers.
-     * This property does not affect method calls nor method references.
+     * This property does not affect method calls nor method references nor record components.
      */
     private List<Integer> memberModifiers = Collections.emptyList();
 
@@ -473,9 +473,11 @@ public final class IllegalTypeCheck extends AbstractCheck {
             case TokenTypes.VARIABLE_DEF:
             case TokenTypes.ANNOTATION_FIELD_DEF:
             case TokenTypes.PATTERN_VARIABLE_DEF:
-            case TokenTypes.RECORD_COMPONENT_DEF:
                 visitVariableDef(ast);
                 break;
+            case TokenTypes.RECORD_COMPONENT_DEF:
+                checkClassName(ast);
+                break;
             case TokenTypes.PARAMETER_DEF:
                 visitParameterDef(ast);
                 break;
@@ -852,7 +854,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
     /**
      * Setter to control whether to check only methods and fields with any of
      * the specified modifiers.
-     * This property does not affect method calls nor method references.
+     * This property does not affect method calls nor method references nor record components.
      *
      * @param modifiers String contains modifiers.
      */
diff --git a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml
index 05b9e707d..51a2ab355 100644
--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml
@@ -80,7 +80,7 @@
                       validation-type="tokenTypesSet">
                <description>Control whether to check only methods and fields with any
  of the specified modifiers.
- This property does not affect method calls nor method references.</description>
+ This property does not affect method calls nor method references nor record components.</description>
             </property>
             <property default-value="ANNOTATION_FIELD_DEF,CLASS_DEF,INTERFACE_DEF,METHOD_CALL,METHOD_DEF,METHOD_REF,PARAMETER_DEF,VARIABLE_DEF,PATTERN_VARIABLE_DEF,RECORD_DEF,RECORD_COMPONENT_DEF"
                       name="tokens"
