diff --git a/src/main/java/com/fasterxml/jackson/databind/deser/std/StdDeserializer.java b/src/main/java/com/fasterxml/jackson/databind/deser/std/StdDeserializer.java
index fa8efe714..76629cb98 100644
--- a/src/main/java/com/fasterxml/jackson/databind/deser/std/StdDeserializer.java
+++ b/src/main/java/com/fasterxml/jackson/databind/deser/std/StdDeserializer.java
@@ -1421,7 +1421,7 @@ public abstract class StdDeserializer<T>
             return ob.toString();
         // 29-Jun-2020, tatu: New! "Scalar from Object" (mostly for XML)
         case JsonTokenId.ID_START_OBJECT:
-            return ctxt.extractScalarFromObject(p, this, _valueClass);
+            return ctxt.extractScalarFromObject(p, this, rawTargetType);
         case JsonTokenId.ID_NUMBER_INT:
             act = _checkIntToStringCoercion(p, ctxt, rawTargetType);
             break;
@@ -1453,7 +1453,7 @@ public abstract class StdDeserializer<T>
                 return text;
             }
         }
-        return (String) ctxt.handleUnexpectedToken(getValueType(ctxt), p);
+        return (String) ctxt.handleUnexpectedToken(rawTargetType, p);
     }
 
     /**
