diff --git a/src/main/java/com/fasterxml/jackson/core/json/UTF8JsonGenerator.java b/src/main/java/com/fasterxml/jackson/core/json/UTF8JsonGenerator.java
index b96e9003d..dbef52e3d 100644
--- a/src/main/java/com/fasterxml/jackson/core/json/UTF8JsonGenerator.java
+++ b/src/main/java/com/fasterxml/jackson/core/json/UTF8JsonGenerator.java
@@ -1344,6 +1344,11 @@ public class UTF8JsonGenerator
 
         while (left > 0) {
             int len = Math.min(_outputMaxContiguous, left);
+            // [core#1473]: avoid splitting surrogates between two segments.
+            // if len == 1 (edge case) don't apply to avoid infinite loop
+            if (len > 1 && _isStartOfSurrogatePair(text.charAt(offset + len-1))) {
+                --len;
+            }
             if ((_outputTail + len) > _outputEnd) { // caller must ensure enough space
                 _flushBuffer();
             }
@@ -1370,6 +1375,11 @@ public class UTF8JsonGenerator
     {
         do {
             int len = Math.min(_outputMaxContiguous, totalLen);
+            // [core#1473]: avoid splitting surrogates between two segments.
+            // if len == 1 (edge case) don't apply to avoid infinite loop
+            if (len > 1 && _isStartOfSurrogatePair(cbuf[offset + len-1])) {
+                --len;
+            }
             if ((_outputTail + len) > _outputEnd) { // caller must ensure enough space
                 _flushBuffer();
             }
@@ -1383,6 +1393,11 @@ public class UTF8JsonGenerator
     {
         do {
             int len = Math.min(_outputMaxContiguous, totalLen);
+            // [core#1473]: avoid splitting surrogates between two segments.
+            // if len == 1 (edge case) don't apply to avoid infinite loop
+            if (len > 1 && _isStartOfSurrogatePair(text.charAt(offset + len-1))) {
+                --len;
+            }
             if ((_outputTail + len) > _outputEnd) { // caller must ensure enough space
                 _flushBuffer();
             }
@@ -1869,6 +1884,11 @@ public class UTF8JsonGenerator
     {
         do {
             int len = Math.min(_outputMaxContiguous, totalLen);
+            // [core#1473]: avoid splitting surrogates between two segments.
+            // if len == 1 (edge case) don't apply to avoid infinite loop
+            if (len > 1 && _isStartOfSurrogatePair(utf8[offset + len-1])) {
+                --len;
+            }
             _writeUTF8Segment(utf8, offset, len);
             offset += len;
             totalLen -= len;
