diff --git a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/model/typesystem/ReferenceTypeImpl.java b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/model/typesystem/ReferenceTypeImpl.java
index 644555627..acad918f8 100644
--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/model/typesystem/ReferenceTypeImpl.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/model/typesystem/ReferenceTypeImpl.java
@@ -117,8 +117,8 @@ public class ReferenceTypeImpl extends ResolvedReferenceType {
         if (other instanceof LambdaArgumentTypePlaceholder) {
             return FunctionalInterfaceLogic.isFunctionalInterfaceType(this);
         }
-        if (other instanceof ReferenceTypeImpl) {
-            ReferenceTypeImpl otherRef = (ReferenceTypeImpl) other;
+        if (other.isReferenceType()) {
+            ResolvedReferenceType otherRef =  other.asReferenceType();
             if (compareConsideringTypeParameters(otherRef)) {
                 return true;
             }
