diff --git a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserClassDeclaration.java b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserClassDeclaration.java
index 4509f4510..aed3a1578 100644
--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserClassDeclaration.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserClassDeclaration.java
@@ -338,7 +338,7 @@ public class JavaParserClassDeclaration extends AbstractClassDeclaration impleme
         if (this.isJavaLangObject()) {
             return ancestors;
         }
-
+        
         Optional<String> qualifiedName = wrappedNode.getFullyQualifiedName();
         if (!qualifiedName.isPresent()) {
             return ancestors;
@@ -389,7 +389,7 @@ public class JavaParserClassDeclaration extends AbstractClassDeclaration impleme
         if (resolvedReferenceTypeDeclaration.isPresent()) {
             ResolvedTypeDeclaration rtd = resolvedReferenceTypeDeclaration.get().asType();
             // do not consider an inner or nested class as an ancestor
-            return !rtd.getQualifiedName().contains(ownQualifiedName);
+            return !rtd.hasInternalType(ownQualifiedName);
         }
         return false;
     }
