diff --git a/brut.apktool/apktool-lib/src/main/java/brut/androlib/AaptInvoker.java b/brut.apktool/apktool-lib/src/main/java/brut/androlib/AaptInvoker.java
index 7aaad3d0..3720829f 100644
--- a/brut.apktool/apktool-lib/src/main/java/brut/androlib/AaptInvoker.java
+++ b/brut.apktool/apktool-lib/src/main/java/brut/androlib/AaptInvoker.java
@@ -182,6 +182,8 @@ public class AaptInvoker {
 
         cmd.add("--allow-reserved-package-id");
 
+        cmd.add("--no-compile-sdk-metadata");
+
         if (mApkInfo.sparseResources) {
             cmd.add("--enable-sparse-encoding");
         }
