diff --git a/tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommadTest.java b/tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommandTest.java
similarity index 95%
rename from tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommadTest.java
rename to tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommandTest.java
index d7315d058..35b0f71ea 100644
--- a/tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommadTest.java
+++ b/tools/src/test/java/org/apache/rocketmq/tools/command/broker/BrokerConsumeStatsSubCommandTest.java
@@ -44,9 +44,9 @@ import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
-public class BrokerConsumeStatsSubCommadTest {
+public class BrokerConsumeStatsSubCommandTest {
 
-    private static BrokerConsumeStatsSubCommad cmd = new BrokerConsumeStatsSubCommad();
+    private static BrokerConsumeStatsSubCommand cmd = new BrokerConsumeStatsSubCommand();
 
     private static DefaultMQAdminExt defaultMQAdminExt;
     private static DefaultMQAdminExtImpl defaultMQAdminExtImpl;
@@ -83,7 +83,7 @@ public class BrokerConsumeStatsSubCommadTest {
     @Test
     public void testExecute() throws SubCommandException, IllegalAccessException, NoSuchFieldException {
 
-        Field field = BrokerConsumeStatsSubCommad.class.getDeclaredField("defaultMQAdminExt");
+        Field field = BrokerConsumeStatsSubCommand.class.getDeclaredField("defaultMQAdminExt");
         field.setAccessible(true);
         field.set(cmd, defaultMQAdminExt);
 
diff --git a/tools/src/test/java/org/apache/rocketmq/tools/command/topic/GetTopicConfigSubCommandTest.java b/tools/src/test/java/org/apache/rocketmq/tools/command/topic/GetTopicConfigSubCommandTest.java
new file mode 100644
index 000000000..3cb64b630
--- /dev/null
+++ b/tools/src/test/java/org/apache/rocketmq/tools/command/topic/GetTopicConfigSubCommandTest.java
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.rocketmq.tools.command.topic;
+
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.Options;
+import org.apache.rocketmq.srvutil.ServerUtil;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GetTopicConfigSubCommandTest {
+    @Test
+    public void testExecute() {
+        GetTopicConfigSubCommand cmd = new GetTopicConfigSubCommand();
+        Options options = ServerUtil.buildCommandlineOptions(new Options());
+        String[] subargs = new String[] {"-b 127.0.0.1:10911", "-t topic-test"};
+        final CommandLine commandLine =
+            ServerUtil.parseCmdLine("mqadmin " + cmd.commandName(), subargs,
+                cmd.buildCommandlineOptions(options), new DefaultParser());
+        assertThat(commandLine.getOptionValue('b').trim()).isEqualTo("127.0.0.1:10911");
+        assertThat(commandLine.getOptionValue("t").trim()).isEqualTo("topic-test");
+    }
+}
