diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
index 3bad20eea..a3a7e5137 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
@@ -81,6 +81,7 @@ import com.puppycrawl.tools.checkstyle.internal.utils.XmlUtil;
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 public class XdocsPagesTest {
+    private static final Path SITE_PATH = Paths.get("src/site/site.xml");
 
     private static final Path AVAILABLE_CHECKS_PATH = Paths.get("src/xdocs/checks.xml");
     private static final String LINK_TEMPLATE =
@@ -243,6 +244,22 @@ public class XdocsPagesTest {
         return availableChecks.matches(linkPattern);
     }
 
+    @Test
+    public void testAllConfigsHaveLinkInSite() throws Exception {
+        final String siteContent = Files.readString(SITE_PATH);
+
+        for (Path path : XdocUtil.getXdocsConfigFilePaths(XdocUtil.getXdocsFilePaths())) {
+            final String expectedFile = path.toString()
+                    .replace(".xml", ".html")
+                    .replaceAll("\\\\", "/")
+                    .replaceAll("src[\\\\/]xdocs[\\\\/]", "");
+            final String expectedLink = String.format(Locale.ROOT, "href=\"%s\"", expectedFile);
+            assertWithMessage("Expected to find link to '" + expectedLink + "' in " + SITE_PATH)
+                    .that(siteContent)
+                    .contains(expectedLink);
+        }
+    }
+
     @Test
     public void testAllChecksPageInSyncWithChecksSummaries() throws Exception {
         final Pattern endOfSentence = Pattern.compile("(.*?\\.)\\s", Pattern.DOTALL);
