diff --git a/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java b/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java
index 0993f39..9e62fbc 100644
--- a/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java
+++ b/common/src/test/java/com/alibaba/nacos/common/remote/client/RpcClientTest.java
@@ -440,6 +440,7 @@ public class RpcClientTest {
         } catch (InvocationTargetException e) {
             e.printStackTrace();
         }
-        verify(connection, times(3)).request(any(), anyLong());
+        // 1 normal call, 3 retries
+        verify(connection, times(4)).request(any(), anyLong());
     }
 }
\ No newline at end of file
