diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheckTest.java
index 8783d4b..e003950 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheckTest.java
@@ -122,6 +122,7 @@ public class ClassMemberImpliedModifierCheckTest
             "29:9: " + getCheckMessage(MSG_KEY, "static"),
             "33:9: " + getCheckMessage(MSG_KEY, "static"),
             "37:9: " + getCheckMessage(MSG_KEY, "static"),
+            "43:13: " + getCheckMessage(MSG_KEY, "static"),
         };
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputClassMemberImpliedModifierRecords.java"),
