diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java
index db40f5613c..81457b7bb1 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LocalVariableNameCheckTest.java
@@ -50,7 +50,7 @@ public class LocalVariableNameCheckTest
     public void testDefault()
             throws Exception {
 
-        final String pattern = "^[a-z][a-zA-Z0-9]*$";
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expected = {
             "122:13: " + getCheckMessage(MSG_INVALID_PATTERN, "ABC", pattern),
@@ -88,4 +88,18 @@ public class LocalVariableNameCheckTest
                 getPath("InputLocalVariableNameOneCharInitVarName.java"), expected);
     }
 
+    @Test
+    public void testUnnamedVariables() throws Exception {
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
+
+        final String[] expected = {
+            "16:13: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "17:13: " + getCheckMessage(MSG_INVALID_PATTERN, "_result", pattern),
+            "33:22: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputLocalVariableNameUnnamedVariables.java"), expected);
+
+    }
+
 }
