diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java
index a3d195300e..57793538cb 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheckTest.java
@@ -58,7 +58,7 @@ public class CatchParameterNameCheckTest extends AbstractModuleTestSupport {
 
     @Test
     public void testDefaultConfigurationOnFileWithViolations() throws Exception {
-        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+)$";
+        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
 
         final String[] expected = {
             "25:28: " + getCheckMessage(MSG_INVALID_PATTERN, "exception1", defaultFormat),
@@ -96,4 +96,17 @@ public class CatchParameterNameCheckTest extends AbstractModuleTestSupport {
                 getPath("InputCatchParameterName3.java"), expected);
     }
 
+    @Test
+    public void testCatchParameterNameUnnamed() throws Exception {
+        final String defaultFormat = "^(e|t|ex|[a-z][a-z][a-zA-Z]+|_)$";
+
+        final String[] expected = {
+            "18:28: " + getCheckMessage(MSG_INVALID_PATTERN, "__", defaultFormat),
+            "24:28: " + getCheckMessage(MSG_INVALID_PATTERN, "_BAD", defaultFormat),
+            "27:28: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD__", defaultFormat),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputCatchParameterNameUnnamed.java"), expected);
+    }
 }
