diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
index 91b9cac..262664f 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
@@ -64,6 +64,13 @@ public class NoWhitespaceAfterCheckTest
     }
 
     @Test
+    public void testAssignment() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getPath("InputNoWhitespaceAfterTestAssignment.java"), expected);
+    }
+
+    @Test
     public void testDotAllowLineBreaks() throws Exception {
         final String[] expected = {
             "9:13: " + getCheckMessage(MSG_KEY, "."),
diff --git a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestAssignment.java b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestAssignment.java
new file mode 100644
index 000000000..de870d857
--- /dev/null
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestAssignment.java
@@ -0,0 +1,41 @@
+/*
+NoWhitespaceAfter
+allowLineBreaks = false
+tokens = (default)ARRAY_INIT, AT, INC, DEC, UNARY_MINUS, UNARY_PLUS, BNOT, LNOT, \
+         DOT, ARRAY_DECLARATOR, INDEX_OP
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespaceafter; // ^ 2 violations above
+
+public class InputNoWhitespaceAfterTestAssignment {
+
+    Object o;
+    static boolean b = true;
+
+    void some() {
+        Object oo = new Object[4];
+        Object[] oo2 = new Object[4];
+        this.o = ((Object[]) oo)[1]; // ok
+        this.o = ((java.lang.Object[]) oo)[1]; // ok
+        this.o = oo2[1];
+        QualifiedAssignment.o1 = ((Object[]) oo)[1]; // ok
+        QualifiedAssignment.o1 = ((java.lang.Object[]) oo)[1]; // ok
+        QualifiedAssignment.o1 = oo2[1];
+        QualifiedAssignment qa1 = null;
+        QualifiedAssignment[] qa2 = null;
+        int idx = 0;
+        (qa1 = (QualifiedAssignment)qa2[idx]).o1 = (new QualifiedAssignment[idx][idx][idx])[idx];
+        (b ? (new QualifiedAssignment().q1 = new QualifiedAssignment()) :
+            (QualifiedAssignment)(new QualifiedAssignment().q1 = new QualifiedAssignment())).q1 =
+            (new QualifiedAssignment[new QualifiedAssignment().idx = (QualifiedAssignment.idx =
+                QualifiedAssignment.idx)])[QualifiedAssignment.idx];
+    }
+}
+
+class QualifiedAssignment {
+    static Object o1;
+    static QualifiedAssignment q1;
+    static int idx = 1;
+}
diff --git a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java
index 005d752ae..ecd13e22e 100644
--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterTestDefault.java
@@ -192,7 +192,7 @@ class InputNoWhitespaceAfterTestDefault
     /** bug 806243 (NoWhitespaceBeforeCheck violation for anonymous inner class) */
     void bug806243()
     {
-        Object o = new InputNoWhitespaceAfterTestDefault() {
+        Object o = new InputNoWhitespaceAfterTestAssignment() {
             private int j ;
             //           ^ whitespace
         };
