diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
index 6140743f5b..f0585b14df 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
@@ -154,6 +154,9 @@ public class RedundantModifierCheckTest
             TokenTypes.RESOURCE,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.PATTERN_VARIABLE_DEF,
+            TokenTypes.LITERAL_CATCH,
+            TokenTypes.LAMBDA,
         };
         assertWithMessage("Invalid acceptable tokens")
             .that(actual)
@@ -422,4 +425,34 @@ public class RedundantModifierCheckTest
                 getNonCompilablePath("InputRedundantModifierStrictfpWithDefaultVersion.java"),
                 expected);
     }
+
+    @Test
+    public void testFinalUnnamedVariablesWithDefaultVersion() throws Exception {
+        final String[] expected = {
+            "18:26: " + getCheckMessage(MSG_KEY, "final"),
+            "24:9: " + getCheckMessage(MSG_KEY, "final"),
+            "34:18: " + getCheckMessage(MSG_KEY, "final"),
+            "44:14: " + getCheckMessage(MSG_KEY, "final"),
+            "51:14: " + getCheckMessage(MSG_KEY, "final"),
+            "54:18: " + getCheckMessage(MSG_KEY, "final"),
+            "65:53: " + getCheckMessage(MSG_KEY, "final"),
+            "69:53: " + getCheckMessage(MSG_KEY, "final"),
+            "69:70: " + getCheckMessage(MSG_KEY, "final"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputRedundantModifierFinalUnnamedVariables.java"),
+                expected);
+    }
+
+    @Test
+    public void testFinalUnnamedVariablesWithOldVersion() throws Exception {
+        final String[] expected = {
+            "40:14: " + getCheckMessage(MSG_KEY, "final"),
+            "47:14: " + getCheckMessage(MSG_KEY, "final"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputRedundantModifierFinalUnnamedVariablesWithOldVersion.java"),
+                expected);
+    }
 }
