diff --git a/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/AnimalDto.java b/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/AnimalDto.java
index 87b60c3cdb2..c17b0385407 100644
--- a/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/AnimalDto.java
+++ b/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/AnimalDto.java
@@ -34,7 +34,6 @@ import jakarta.annotation.Generated;
   @JsonSubTypes.Type(value = DogDto.class, name = "Dog")
 })
 
-@JsonTypeName("Animal")
 @Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
 public class AnimalDto {
 
diff --git a/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/CatDto.java b/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/CatDto.java
index 99cf34ad232..61f8bc958e6 100644
--- a/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/CatDto.java
+++ b/samples/client/petstore/spring-http-interface/src/main/java/org/openapitools/model/CatDto.java
@@ -31,7 +31,6 @@ import jakarta.annotation.Generated;
   @JsonSubTypes.Type(value = BigCatDto.class, name = "BigCat")
 })
 
-@JsonTypeName("Cat")
 @Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
 public class CatDto extends AnimalDto {
 
diff --git a/samples/server/petstore/springboot/src/main/java/org/openapitools/model/AnimalDto.java b/samples/server/petstore/springboot/src/main/java/org/openapitools/model/AnimalDto.java
index aba7e9e70f5..3d6817c8f8c 100644
--- a/samples/server/petstore/springboot/src/main/java/org/openapitools/model/AnimalDto.java
+++ b/samples/server/petstore/springboot/src/main/java/org/openapitools/model/AnimalDto.java
@@ -37,7 +37,6 @@ import javax.annotation.Generated;
   @JsonSubTypes.Type(value = DogDto.class, name = "Dog")
 })
 
-@JsonTypeName("Animal")
 @Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
 public class AnimalDto {
 
diff --git a/samples/server/petstore/springboot/src/main/java/org/openapitools/model/CatDto.java b/samples/server/petstore/springboot/src/main/java/org/openapitools/model/CatDto.java
index ffbd4428513..3d11bd7ba3a 100644
--- a/samples/server/petstore/springboot/src/main/java/org/openapitools/model/CatDto.java
+++ b/samples/server/petstore/springboot/src/main/java/org/openapitools/model/CatDto.java
@@ -34,7 +34,6 @@ import javax.annotation.Generated;
   @JsonSubTypes.Type(value = BigCatDto.class, name = "BigCat")
 })
 
-@JsonTypeName("Cat")
 @Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
 public class CatDto extends AnimalDto {
 
