diff --git a/samples/client/petstore/java/okhttp-gson-awsv4signature/src/main/java/org/openapitools/client/model/Pet.java b/samples/client/petstore/java/okhttp-gson-awsv4signature/src/main/java/org/openapitools/client/model/Pet.java
index e07848f761c..649c81edcba 100644
--- a/samples/client/petstore/java/okhttp-gson-awsv4signature/src/main/java/org/openapitools/client/model/Pet.java
+++ b/samples/client/petstore/java/okhttp-gson-awsv4signature/src/main/java/org/openapitools/client/model/Pet.java
@@ -122,6 +122,7 @@ public class Pet {
   }
 
   public static final String SERIALIZED_NAME_STATUS = "status";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_STATUS)
   private StatusEnum status;
 
@@ -254,6 +255,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public Pet status(StatusEnum status) {
     
     this.status = status;
@@ -273,6 +275,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public void setStatus(StatusEnum status) {
     this.status = status;
   }
diff --git a/samples/client/petstore/java/okhttp-gson-group-parameter/src/main/java/org/openapitools/client/model/Pet.java b/samples/client/petstore/java/okhttp-gson-group-parameter/src/main/java/org/openapitools/client/model/Pet.java
index e07848f761c..649c81edcba 100644
--- a/samples/client/petstore/java/okhttp-gson-group-parameter/src/main/java/org/openapitools/client/model/Pet.java
+++ b/samples/client/petstore/java/okhttp-gson-group-parameter/src/main/java/org/openapitools/client/model/Pet.java
@@ -122,6 +122,7 @@ public class Pet {
   }
 
   public static final String SERIALIZED_NAME_STATUS = "status";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_STATUS)
   private StatusEnum status;
 
@@ -254,6 +255,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public Pet status(StatusEnum status) {
     
     this.status = status;
@@ -273,6 +275,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public void setStatus(StatusEnum status) {
     this.status = status;
   }
diff --git a/samples/client/petstore/java/okhttp-gson-swagger1/src/main/java/org/openapitools/client/model/Pet.java b/samples/client/petstore/java/okhttp-gson-swagger1/src/main/java/org/openapitools/client/model/Pet.java
index f830ef3541c..1e9e75d54c7 100644
--- a/samples/client/petstore/java/okhttp-gson-swagger1/src/main/java/org/openapitools/client/model/Pet.java
+++ b/samples/client/petstore/java/okhttp-gson-swagger1/src/main/java/org/openapitools/client/model/Pet.java
@@ -125,6 +125,7 @@ public class Pet {
   }
 
   public static final String SERIALIZED_NAME_STATUS = "status";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_STATUS)
   private StatusEnum status;
 
@@ -262,6 +263,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public Pet status(StatusEnum status) {
     
     this.status = status;
@@ -282,6 +284,7 @@ public class Pet {
   }
 
 
+  @Deprecated
   public void setStatus(StatusEnum status) {
     this.status = status;
   }
diff --git a/samples/client/petstore/java/okhttp-gson/src/main/java/org/openapitools/client/model/ObjectWithDeprecatedFields.java b/samples/client/petstore/java/okhttp-gson/src/main/java/org/openapitools/client/model/ObjectWithDeprecatedFields.java
index c7b55901d67..563308ff78a 100644
--- a/samples/client/petstore/java/okhttp-gson/src/main/java/org/openapitools/client/model/ObjectWithDeprecatedFields.java
+++ b/samples/client/petstore/java/okhttp-gson/src/main/java/org/openapitools/client/model/ObjectWithDeprecatedFields.java
@@ -57,14 +57,17 @@ public class ObjectWithDeprecatedFields {
   private String uuid;
 
   public static final String SERIALIZED_NAME_ID = "id";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_ID)
   private BigDecimal id;
 
   public static final String SERIALIZED_NAME_DEPRECATED_REF = "deprecatedRef";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_DEPRECATED_REF)
   private DeprecatedObject deprecatedRef;
 
   public static final String SERIALIZED_NAME_BARS = "bars";
+  @Deprecated
   @SerializedName(SERIALIZED_NAME_BARS)
   private List<String> bars;
 
@@ -93,6 +96,7 @@ public class ObjectWithDeprecatedFields {
   }
 
 
+  @Deprecated
   public ObjectWithDeprecatedFields id(BigDecimal id) {
     
     this.id = id;
@@ -112,11 +116,13 @@ public class ObjectWithDeprecatedFields {
   }
 
 
+  @Deprecated
   public void setId(BigDecimal id) {
     this.id = id;
   }
 
 
+  @Deprecated
   public ObjectWithDeprecatedFields deprecatedRef(DeprecatedObject deprecatedRef) {
     
     this.deprecatedRef = deprecatedRef;
@@ -136,11 +142,13 @@ public class ObjectWithDeprecatedFields {
   }
 
 
+  @Deprecated
   public void setDeprecatedRef(DeprecatedObject deprecatedRef) {
     this.deprecatedRef = deprecatedRef;
   }
 
 
+  @Deprecated
   public ObjectWithDeprecatedFields bars(List<String> bars) {
     
     this.bars = bars;
@@ -168,6 +176,7 @@ public class ObjectWithDeprecatedFields {
   }
 
 
+  @Deprecated
   public void setBars(List<String> bars) {
     this.bars = bars;
   }
